### ISSUE TITLE
TFRecord integration with Keras API

### ISSUE THREAD

**Issue by ahundt** [tensorflow/tensorflow#8787]
Since the Keras-2 API is now directly in TensorFlow, I think it would be very useful if there were a mechanism directly in TensorFlow to supply TFRecords to Keras, such as in a call to `model.fit()`, or with equivalent functionality to `flow_from_directory`.

One key implementation detail is with the way `model.compile()` works as [detailed in a comment on this topic in keras](https://github.com/fchollet/keras/issues/5358#issuecomment-282117525). As discussed in the same comment, numpy arrays have been determined to be the primary way of interacting with Keras, so this should not be an upstream Keras request because TFRecords currently require TensorFlow, so other backends couldn't be supported easily.

New Pull Request (includes examples):
https://github.com/fchollet/keras/pull/6928

(outdated) Starter code:
https://github.com/farizrahman4u/keras-contrib/pull/27

Related Keras issues:
https://github.com/fchollet/keras/issues/5356
https://github.com/fchollet/keras/issues/5368
https://github.com/fchollet/keras/issues/5325

**Comment by terrytangyuan** [comment]
Since Keras will work with Estimator API, we can put this integration in `tf.estimator.inputs`? I think you can actually just ingest your data from TFRecords source to as part of `input_fn` in Estimator API. Similarly, this will work with `Experiment` as well.

**Comment by ahundt** [comment]
@terrytangyuan that sounds good, is what you describe forward looking or already integrated?

**Comment by terrytangyuan** [comment]
I don't think it's integrated yet since Keras literally just landed in contrib. I am not in the internal team but @martinwicke knows the plan of things.

**Comment by martinwicke** [comment]
Keras just landed in contrib and will undergo some internal refactoring for a bit. The plan is to provide simple ways to:

- use Keras inside a model_fn (already the case, Keras layers work fine inside Estimators)
- use a compiled Keras Model to create an Estimator
- add the ability for a Keras Model to accept an input_fn for its fit() method.

This will take a while but it's on the roadmap (@fchollet you may correct me on the last bullet).

**Comment by ahundt** [comment]
Thanks I saw it was just added, just wasn't sure of the status.

Using `Experiment` with Keras models does sound good. The other way around may also be useful as an option, where if you already have training code for `Model.fit()` and an existing TFRecord dataset.

**Comment by martinwicke** [comment]
Yes, that would be the "allowing input_fn for Model".

**Comment by ahundt** [comment]
Sounds great! Thanks for clarifying, would it be good to leave this open to follow the progress?

**Comment by martinwicke** [comment]
Yes, let's leave this open.

**Comment by ahundt** [comment]
I'm learning as I go but I've been looking into this a bit and here are my first thoughts.

These are from upstream keras, but these two commits/prs take care of a couple of the key parameterizations necessary for TFRecords to work:
- [tensors feeding models](https://github.com/fchollet/keras/commit/e90b0713f2e86f9f540b57b4306a77e3f696213b) - numpy arrays no longer required, merged some time ago
- [kwargs passed to session.run](https://github.com/fchollet/keras/pull/6693/files) - in progress at time of writing

One possible design is to create a `TFRecordInput()` Keras layer that wraps `data_flow_ops.RecordInput` which might be able to fit well with the [tf_cnn_benchmarks suggested high performance design](https://github.com/tensorflow/benchmarks/blob/master/scripts/tf_cnn_benchmarks/preprocessing.py#L360). Then calls like `model.fit()` could get their data from those input tensors.

**Two possible cases I've considered for loading tensors**

`Input()` or `TFRecordInput()` case:
Something like [these checks in the densenet model for the variable input_tensor](https://github.com/farizrahman4u/keras-contrib/blob/master/keras_contrib/applications/densenet.py#L126) might be possible to fold into the Input, or to put on a backend. 

`model.fit()` case:
`model.fit` is after compile which instantiates the tf graph so this might be untenable. Nonetheless, in an ideal world one could create `data_flow_ops.RecordInput()`, loading/preprocessing ops, then those tensors could be fed to `model.fit()`, which could check the input type, and then the tensors could be passed through the keras backend to the TensorFlow backend.

**Preprocessing Layers**

TensorFlow has preprocessing layers and it might be possible to simply use them to create [Keras Preprocessing/Augmentation Layers](https://github.com/fchollet/keras/issues/6655). However there would ideally be some mechanism by which they can be separated from the Keras model when desired.

Hopefully none of this is too far behind tf's ongoing changes, any suggestions/feedback?

**Comment by Netzeband** [comment]
I'm also very interested in using TF Record-files with Keras. I read several issues, that are related to this feature, read the stack-overflow question (https://stackoverflow.com/questions/42184863/how-do-you-make-tensorflow-keras-fast-with-a-tfrecord-dataset) and saw the pull request https://github.com/farizrahman4u/keras-contrib/pull/27 

What is the current state of this Feature? 
Can I merge the pull request into my local repository, or will the final implementation very different to this?

**Comment by ahundt** [comment]
It will be very different. Implementation has not yet started.

**Comment by Netzeband** [comment]
Oh, too bad. When reading the stackoverflow answer, I really thought, that this feature is almost done :-) 

I was thinking about using Keras (the first time) for my next project. But since I have over 100GB TFRecord-Files as training data, this seems not to be an good option currently. 

However thank you very much for driving this topic. If there was a prototype implementation, I would have tested it with pleasure. But for a full implementation, I think I do not have enough knowledge about Keras at the moment. So I cannot help much. Thus I will check again in some month, if your ideas are implemented. Keep going 👍

**Comment by normanheckscher** [comment]
I'm in a similar situation to you, netzeband. 

I'm sure Andrew will correct me if I'm wrong, however, the code in the PR looks good as a single example on 28x28 single channel images. It doesn't appear that it will work with other datasets unless you change the code in the PR to suit your own dataset. For the code to be pulled into TensorFlow master, it'd need to be genericised and also be ported into Keras for use with Theano, so as not to fork the original Keras code base. 

I'm looking at using some of this PR code and modifying it to work with my data, not as part of TensorFlow, but as part of my own project. Unless there's some other way?

**Comment by ahundt** [comment]
@normanheckscher is correct

**Comment by drewrice2** [comment]
here to show interest for this feature. thanks @ahundt for fighting the good fight!

**Comment by ahundt** [comment]
Notable progress in upstream Keras https://github.com/fchollet/keras/pull/6928 for `model.fit()`. Right now there are a few bugs caught by unit tests so I'd appreciate comments/suggestions.

**Comment by ahundt** [comment]
https://github.com/fchollet/keras/pull/6928 unit tests pass

**Comment by martinwicke** [comment]
So, what should we do with this issue? Is this relevant or is it covered by recent pull from Keras? @fchollet?

**Comment by ahundt** [comment]
@martinwicke fchollet/keras#6928 integrates TFRecords via yield op support (i.e. [RecordInput](https://github.com/tensorflow/tensorflow/blob/833252af72af56661aefb0541163109132f9d4a6/tensorflow/python/ops/data_flow_ops.py#L2137)). The fifo queue way to get records and the new Dataset class may or may not work with Keras after those changes, but if they do an example is probably a good idea.

I'd suggest not closing this until the changes are merged to TensorFlow from upstream Keras. If the other kinds of data input should also be supported this issue might still apply, but a new issue may also be appropriate.

**Comment by ahundt** [comment]
Based on the discussion in [Keras Input Tensor API Design Proposal](https://docs.google.com/document/d/1tf2Nl7wor8rmWPUoxfClLuPLQGqvZryegD7K7-1tTe8/edit?usp=sharing) and https://github.com/fchollet/keras/pull/6928#issuecomment-312088200 implementation of https://github.com/tensorflow/tensorflow/issues/10837 will be required for the ideal API design that supports TFRecords (and any data input op) to work.

**Comment by ahundt** [comment]
https://github.com/fchollet/keras/issues/7503 has the latest action plan, pieces of this functionality are being merged into keras upstream

**Comment by bhack** [comment]
Will be also an interoperability also with `tf.contrib.data.Dataset`?

**Comment by fchollet** [comment]
Yes, you can use the Dataset API to train Keras models. The standard
tf.keras workflow would likely involve managing data via the Dataset API,
building your model via the Keras API, and training it via the Estimator
API.

**Comment by dartdog** [comment]
Interested also just found this when trying to move a tf.estimator based model to Keras and realizing the data prep and

`training_input_fn = tf.estimator.inputs.pandas_input_fn(x=training_data, y=training_label, batch_size=64, shuffle=True, num_epochs=None)`

Would not work to feed Keras, looking forward to it landing, the ability to use the estimator and experiment stuff with Keras is interesting

**Comment by babakEhteshami** [comment]
Checking in to announce my interest as well!

**Comment by fchollet** [comment]
The `estimator = model.get_estimator()` feature is now available for `tf.keras` (in TF 1.4).

To train a Keras model from TFRecords, please see this example: https://github.com/fchollet/keras/blob/master/examples/mnist_tfrecord.py

**Comment by mike-grayhat** [comment]
Are there any disadvantages in converting input_tensor from tfrecords into a generator this way:

```
def` tensors2gen(*input_tensors):
    while True:
        data = sess.run(input_tensors)
        if len(input_tensors) == 1:
            data = data[0]
        yield data
```

and then train a model using `fit_generator` that also supports validation data parameter?

**Comment by mike-grayhat** [comment]
Thank you for the clarification!

I played wit tf estimators a bit. I have to train a stacked denoising autoencoder model - so i have to stack layers on top of the existed models and seems like there is no easy way to do this with estimators comparing to the simplicity of keras. And i also didn't find a good guide how to do validation during training phase with estimators (there are some complaints about weights reinitializations on test/validation switches). So i think i will stick with generators for now because i don't care much about performance at this point.

**Comment by siddsuresh97** [comment]
Has anyone tried using tf-records to train a keras model with multiple inputs?
